mirgecom/restart.py

```diff
@@ -0,0 +1,68 @@
+"""Provide some utilities for restarting simulations.
+
+.. autofunction:: read_restart_data
+.. autofunction:: write_restart_file
+.. autofunction:: make_fluid_state
+
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import pickle
+from meshmode.dof_array import unflatten, flatten
+from mirgecom.fluid import make_conserved
+
+
+def read_restart_data(filename):
+    """Read the raw restart data dictionary from the given pickle restart file."""
+    with open(filename, "rb") as f:
+        restart_data = pickle.load(f)
+    return restart_data
+
+
+def make_fluid_state(actx, discr, restart_q):
+    """Make a :class:`~mirgecom.fluid.ConservedVars` from pickled restart data."""
+    from pytools.obj_array import obj_array_vectorize
+    q = unflatten(actx, discr, obj_array_vectorize(actx.from_numpy, restart_q))
+    return make_conserved(discr.dim, scalar_quantities=q)
```

Comment:
As below: Not sure this is needed. Using [this context manager](https://documen.tician.de/meshmode/array.html?highlight=array_context_for#meshmode.dof_array.array_context_for_pickling) (https://github.com/inducer/meshmode/pull/229 has some doc improvements), `ConservedVars` itself should be picklable, cf. #188.